File: ios/Plugin/AudioAsset.swift
//
//  AudioAsset.swift
//  Plugin
//
//  Created by priyank on 2020-05-29.
//  Copyright © 2022 Martin Donadieu. All rights reserved.
//

import Foundation
import AVFoundation
import os.log

/**
 * AudioAsset class handles local audio playback via AVAudioPlayer
 * Supports volume control, fade effects, rate changes, and looping
 */
public class AudioAsset: NSObject, AVAudioPlayerDelegate {

    open var channels: [AVAudioPlayer] = []
    open var playIndex: Int = 0
    open var assetId: String = ""
    open var initialVolume: Float = 1.0
    public let zeroVolume: Float = 0.001 // Minimum volume to avoid zero for exponential fade
    public let maxVolume: Float = 1.0
    open weak var owner: NativeAudio?
    private var logger = Logger(logTag: "AudioAsset")

    // Constants for fade effect
    public let fadeDelaySecs: Float = 0.08

    open var currentTimeTimer: Timer?
    open var fadeTimer: Timer?

    open var fadeTask: DispatchWorkItem?
    public let fadeQueue: DispatchQueue = DispatchQueue(label: "com.audioasset.fadeQueue")

    open var dispatchedCompleteMap: [String: Bool] = [:]

    /**
     * Initialize a new audio asset
     * - Parameters:
     *   - owner: The plugin that owns this asset
     *   - assetId: Unique identifier for this asset
     *   - path: File path to the audio file
     *   - channels: Number of simultaneous playback channels (polyphony)
     *   - volume: Initial volume (0.0-1.0)
     */
    init(owner: NativeAudio, withAssetId assetId: String, withPath path: String!, withChannels channels: Int!, withVolume volume: Float!) {

        self.owner = owner
        self.assetId = assetId
        self.channels = []
        self.initialVolume = min(max(volume ?? Constant.DefaultVolume, Constant.MinVolume), Constant.MaxVolume) // Validate volume range

        super.init()

        guard let encodedPath = path.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
            logger.error("Failed to encode path: %@", String(describing: path))
            return
        }

        // Try to create URL from string first, fall back to file URL if that fails
        let pathUrl: URL
        if let url = URL(string: encodedPath) {
            pathUrl = url
        } else {
            pathUrl = URL(fileURLWithPath: encodedPath)
        }

        // Limit channels to a reasonable maximum to prevent resource issues
        let channelCount = min(max(channels ?? 1, 1), Constant.MaxChannels)
        
        // Create the players directly on the current queue when in test mode
        // to avoid potential deadlocks
        let setupBlock = { [weak self] in
            guard let self = self else { return }
            for _ in 0..<channelCount {
                do {
                    let player = try AVAudioPlayer(contentsOf: pathUrl)
                    player.delegate = self
                    player.enableRate = true
                    player.volume = self.initialVolume
                    player.rate = 1.0
                    player.prepareToPlay()
                    self.channels.append(player)
                } catch {
                    logger.error("Error loading audio file: %@ - path: %@", error.localizedDescription, String(describing: path))
                }
            }
        }
        
        // In test mode, run setup directly to avoid deadlock
        if owner.isRunningTests {
            setupBlock()
        } else {
            owner.executeOnAudioQueue(setupBlock)
        }
    }

    deinit {
        stopCurrentTimeUpdates()
        stopFadeTimer()
        cancelFade()
        for player in channels {
            if player.isPlaying {
                player.stop()
            }
        }
        channels.removeAll()
        dispatchedCompleteMap.removeAll()
    }

    /**
     * Get the current playback time asynchronously
     * - Parameter callback: Closure to receive current time in seconds
     */
    func getCurrentTime(callback: @escaping (TimeInterval) -> Void) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { callback(0); return }
            if channels.isEmpty || playIndex >= channels.count {
                callback(0)
                return
            }
            let player = channels[playIndex]
            callback(player.currentTime)
        }
    }

    /**
     * Set the current playback time
     * - Parameter time: Time in seconds
     */
    func setCurrentTime(time: TimeInterval) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            if channels.isEmpty || playIndex >= channels.count {
                return
            }
            let player = channels[playIndex]
            // Ensure time is valid
            let validTime = min(max(time, 0), player.duration)
            player.currentTime = validTime
        }
    }

    /**
     * Get the total duration of the audio file asynchronously
     * - Parameter callback: Closure to receive duration in seconds
     */
    func getDuration(callback: @escaping (TimeInterval) -> Void) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { callback(0); return }
            if channels.isEmpty || playIndex >= channels.count {
                callback(0)
                return
            }
            let player = channels[playIndex]
            callback(player.duration)
        }
    }

    /**
     * Play the audio from the specified time with optional delay
     * - Parameters:
     *   - time: Start time in seconds
     *   - volume: Volume level (0.0-1.0)
     */
    func play(time: TimeInterval, volume: Float? = nil) {
        stopCurrentTimeUpdates()
        stopFadeTimer()

        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            guard !channels.isEmpty else { return }

            // Reset play index if it's out of bounds
            if playIndex >= channels.count {
                playIndex = 0
            }

            // Ensure the audio session is active before playing
            owner?.activateSession()

            let player = channels[playIndex]
            // Ensure time is within valid range
            let validTime = min(max(time, 0), player.duration)
            player.currentTime = validTime
            player.numberOfLoops = 0
            player.volume = volume ?? self.initialVolume
            player.play()

            playIndex = (playIndex + 1) % channels.count
            startCurrentTimeUpdates()
        }
    }

    /**
     * Play the audio with fade-in effect
     * - Parameters:
     *   - time: Start time in seconds
     *   - volume: Volume level (0.0-1.0)
     *   - fadeInDuration: Duration of the fade-in effect in seconds
     */
    func playWithFade(time: TimeInterval, volume: Float?, fadeInDuration: TimeInterval) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            guard !channels.isEmpty else { return }

            if playIndex >= channels.count {
                playIndex = 0
            }

            let player = channels[playIndex]
            player.currentTime = time

            if !player.isPlaying {
                player.numberOfLoops = 0
                player.volume = 0
                player.play()
                playIndex = (playIndex + 1) % channels.count
                startCurrentTimeUpdates()

                self.fadeIn(audio: player, fadeInDuration: fadeInDuration, targetVolume: volume ?? self.initialVolume)
            }
        }
    }

    func fadeIn(audio: AVAudioPlayer, fadeInDuration: TimeInterval, targetVolume: Float) {
        cancelFade() // Cancel any ongoing fade
        let steps = Int(fadeInDuration / TimeInterval(fadeDelaySecs))
        guard steps > 0 else { return }
        let startVolume: Float = audio.volume
        let fadeStep = (targetVolume - startVolume) / Float(steps)
        var currentVolume: Float = startVolume

        getCurrentTime { startTime in
            self.logger.debug("Beginning fade in at time %2f over @%2f seconds to target volume %2f in %d steps (step duration: %2fs)", startTime, fadeInDuration, targetVolume, steps, self.fadeDelaySecs)
        }
        var task: DispatchWorkItem!
        task = DispatchWorkItem { [weak self] in
            guard !task.isCancelled else { return }
            self?.isPlaying { isPlaying in
                guard let strongSelf = self, isPlaying, audio.isPlaying else {
                    task.cancel()
                    return
                }
                for _ in 0..<steps {
                    let previousCurrentVolume = currentVolume
                    currentVolume += fadeStep
                    let thisTargetVolume = min(max(currentVolume, 0), targetVolume)
                    DispatchQueue.main.async {
                        self?.isPlaying { isPlaying in
                            guard let strongerSelf = self, isPlaying, audio.isPlaying else {
                                task.cancel()
                                return
                            }
                            strongerSelf.logger.verbose("Fade in step: from %2f to %2f to target %2f", previousCurrentVolume, currentVolume, thisTargetVolume)
                            audio.volume = thisTargetVolume
                        }
                    }
                    Thread.sleep(forTimeInterval: TimeInterval(strongSelf.fadeDelaySecs))
                }
                strongSelf.getCurrentTime { endTime in
                    strongSelf.logger.debug("Fade in complete at time %2f", endTime)
                }
            }
        }
        fadeTask = task
        fadeQueue.async(execute: task)
    }

    func fadeOut(audio: AVAudioPlayer, fadeOutDuration: TimeInterval, toPause: Bool = false) {
        cancelFade() // Cancel any ongoing fade
        let steps = Int(fadeOutDuration / TimeInterval(fadeDelaySecs))
        guard steps > 0 else { return }
        var currentVolume: Float = audio.volume
        let fadeStep = currentVolume / Float(steps)

        getCurrentTime { startTime in
            self.logger.debug("Beginning fade out from volume %2f at time %2f over @%2f seconds in %d steps (step duration: %2fs)", currentVolume, startTime, fadeOutDuration, steps, self.fadeDelaySecs)
        }
        var task: DispatchWorkItem!
        task = DispatchWorkItem { [weak self] in
            for _ in 0..<steps {
                guard !task.isCancelled else { return }
                self?.isPlaying { isPlaying in
                    guard let strongSelf = self, isPlaying, audio.isPlaying else {
                        task.cancel()
                        return
                    }
                    let previousCurrentVolume = currentVolume
                    currentVolume -= fadeStep
                    DispatchQueue.main.async {
                        self?.isPlaying { isPlaying in
                            guard let strongerSelf = self, isPlaying, audio.isPlaying else {
                                task.cancel()
                                return
                            }
                            let thisTargetVolume = max(currentVolume, 0)
                            strongerSelf.logger.verbose("Fade out step: from %2f to %2f to target %2f", previousCurrentVolume, currentVolume, thisTargetVolume)
                            audio.volume = thisTargetVolume
                        }
                    }
                    Thread.sleep(forTimeInterval: TimeInterval(strongSelf.fadeDelaySecs))
                }
            }
            DispatchQueue.main.async { [weak self] in
                self?.isPlaying { isPlaying in
                    guard let strongSelf = self, isPlaying, audio.isPlaying else {
                        return
                    }
                    if toPause {
                        audio.pause()
                    } else {
                        audio.stop()
                        strongSelf.owner?.notifyListeners("complete", data: [
                            "assetId": strongSelf.assetId as Any
                        ])
                        strongSelf.dispatchedCompleteMap[strongSelf.assetId] = true
                    }
                    strongSelf.getCurrentTime { endTime in
                        strongSelf.logger.debug("Fade out complete at time %2f", endTime)
                    }
                }
            }
        }
        fadeTask = task
        fadeQueue.async(execute: task)
    }

    func fadeTo(audio: AVAudioPlayer, fadeDuration: TimeInterval, targetVolume: Float) {
        cancelFade() // Cancel any ongoing fade

        let steps = Int(fadeDuration / TimeInterval(fadeDelaySecs))
        guard steps > 0 else { return }

        let minVolume = zeroVolume
        var currentVolume: Float = max(audio.volume, minVolume)
        let safeTargetVolume: Float = max(targetVolume, minVolume)

        // Calculate the exponential ratio
        let ratio = pow(safeTargetVolume / currentVolume, 1.0 / Float(steps))

        getCurrentTime { startTime in
            self.logger.debug("Beginning exponential fade from volume %2f to %2f at time %2f over %2f seconds in %d steps (step duration: %2fs)", currentVolume, safeTargetVolume, startTime, fadeDuration, steps, self.fadeDelaySecs)
        }

        var task: DispatchWorkItem!
        task = DispatchWorkItem { [weak self] in
            guard !task.isCancelled else { return }
            self?.isPlaying { isPlaying in
                guard let strongSelf = self, isPlaying, audio.isPlaying else {
                    task.cancel()
                    return
                }
                for _ in 0..<steps {
                    let previousCurrentVolume = currentVolume
                    currentVolume *= ratio
                    DispatchQueue.main.async {
                        self?.isPlaying { isPlaying in
                            guard let strongerSelf = self, isPlaying, audio.isPlaying else {
                                task.cancel()
                                return
                            }
                            let thisTargetVolume = min(max(currentVolume, minVolume), strongerSelf.maxVolume)
                            strongerSelf.logger.verbose("Exponential fade step: from %2f to %2f to target %2f", previousCurrentVolume, currentVolume, thisTargetVolume)
                            audio.volume = thisTargetVolume
                        }
                    }
                    Thread.sleep(forTimeInterval: TimeInterval(strongSelf.fadeDelaySecs))
                }
                strongSelf.getCurrentTime { endTime in
                    strongSelf.logger.debug("Exponential fade complete at time %2f", endTime)
                }
            }
        }
        fadeTask = task
        fadeQueue.async(execute: task)
    }

    open func cancelFade() {
        if let task = fadeTask {
            task.cancel()
        }
        fadeTask = nil
    }

    internal func stopFadeTimer() {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            if let timer = self.fadeTimer {
                timer.invalidate()
                self.fadeTimer = nil
            }
        }
    }

    open func pause() {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            cancelFade()
            stopCurrentTimeUpdates()

            // Check for valid playIndex
            guard !channels.isEmpty && playIndex < channels.count else { return }

            let player = channels[playIndex]
            player.pause()
        }
    }

    open func resume() {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            // Check for valid playIndex
            guard !channels.isEmpty && playIndex < channels.count else { return }

            let player = channels[playIndex]
            let timeOffset = player.deviceCurrentTime + 0.01
            player.play(atTime: timeOffset)
            startCurrentTimeUpdates()
        }
    }

    open func stop() {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            cancelFade()
            stopCurrentTimeUpdates()
            stopFadeTimer()

            for player in channels {
                if player.isPlaying {
                    player.stop()
                }
                player.currentTime = 0
                player.numberOfLoops = 0
            }
            playIndex = 0

            self.owner?.notifyListeners("complete", data: [
                "assetId": self.assetId
            ])
            self.dispatchedCompleteMap[self.assetId] = true
        }
    }

    open func stopWithFade(fadeOutDuration: TimeInterval, toPause: Bool = false) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            guard !channels.isEmpty && playIndex < channels.count else {
                if !toPause {
                    stop()
                }
                return
            }

            let player = channels[playIndex]
            if player.isPlaying && player.volume > 0 {
                self.fadeOut(audio: player, fadeOutDuration: fadeOutDuration, toPause: toPause)
            } else {
                if !toPause {
                    stop()
                }
            }
        }
    }

    open func loop() {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            self.stop()

            guard !channels.isEmpty && playIndex < channels.count else { return }

            let player = channels[playIndex]
            player.delegate = self
            player.numberOfLoops = -1
            player.play()
            playIndex = (playIndex + 1) % channels.count
            startCurrentTimeUpdates()
        }
    }

    open func unload() {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            cancelFade()
            self.stop()
            stopCurrentTimeUpdates()
            stopFadeTimer()
            channels.removeAll()
            dispatchedCompleteMap.removeAll()
        }
    }

    /**
     * Set the volume for all audio channels
     * - Parameter volume: Volume level (0.0-1.0)
     */
    open func setVolume(volume: NSNumber!, fadeDuration: Double) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            cancelFade()
            // Ensure volume is in valid range
            let validVolume = min(max(volume.floatValue, Constant.MinVolume), Constant.MaxVolume)
            for player in channels {
                if player.isPlaying && fadeDuration > 0 {
                    self.logger.debug("Fade to volume %2f over @%2f seconds", validVolume, fadeDuration)
                    self.fadeTo(audio: player, fadeDuration: fadeDuration, targetVolume: validVolume)
                } else {
                    self.logger.debug("Set volume to %2f", validVolume)
                    player.volume = validVolume
                }
            }
        }
    }

    /**
     * Set the playback rate for all audio channels
     * - Parameter rate: Playback rate (0.5-2.0 is typical range)
     */
    open func setRate(rate: NSNumber!) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            // Ensure rate is in valid range
            let validRate = min(max(rate.floatValue, Constant.MinRate), Constant.MaxRate)
            for player in channels {
                player.rate = validRate
            }
        }
    }

    /**
     * AVAudioPlayerDelegate method called when playback finishes
     */
    public func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { return }

            self.owner?.notifyListeners("complete", data: [
                "assetId": self.assetId
            ])
            self.dispatchedCompleteMap[self.assetId] = true

            // Notify the owner that this player finished
            // The owner will check if any other assets are still playing
            owner?.audioPlayerDidFinishPlaying(player, successfully: flag)
        }
    }

    func playerDecodeError(player: AVAudioPlayer!, error: NSError!) {
        if let error = error {
            logger.error("AudioAsset decode error: %@", error.localizedDescription)
        }
    }

    /**
     * Check if the audio is playing asynchronously
     * - Parameter callback: Closure to receive playing state (Bool)
     */
    open func isPlaying(callback: @escaping (Bool) -> Void) {
        owner?.executeOnAudioQueue { [weak self] in
            guard let self = self else { callback(false); return }
            if channels.isEmpty || playIndex >= channels.count {
                callback(false)
                return
            }
            let player = channels[playIndex]
            callback(player.isPlaying)
        }
    }

    open func startCurrentTimeUpdates() {
        self.stopCurrentTimeUpdates() // Ensure no duplicate timers
        self.dispatchedCompleteMap[self.assetId] = false
        DispatchQueue.main.async { [weak self] in
            guard let strongSelf = self else { return }

            let timer = Timer(timeInterval: 0.1, repeats: true) { [weak self] _ in
                guard let strongSelf = self, let strongOwner = strongSelf.owner else {
                    strongSelf.stopCurrentTimeUpdates()
                    return
                }
                strongSelf.isPlaying { isPlaying in
                    if isPlaying {
                        strongOwner.notifyCurrentTime(strongSelf)
                    } else {
                        strongSelf.stopCurrentTimeUpdates()
                    }
                }
            }
            strongSelf.currentTimeTimer = timer
            RunLoop.current.add(timer, forMode: .common)
        }
    }

    open func stopCurrentTimeUpdates() {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            logger.debug("Stop current time updates")
            if let timer = self.currentTimeTimer {
                timer.invalidate()
                self.currentTimeTimer = nil
            }
        }
    }
}

File: android/src/main/java/ee/forgr/audio/AudioAsset.java
package ee.forgr.audio;

import android.content.res.AssetFileDescriptor;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import androidx.annotation.RequiresApi;
import androidx.media3.common.util.UnstableApi;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

@UnstableApi
public class AudioAsset {

    public static final double DEFAULT_FADE_DURATION_MS = 1000.0;

    private final String TAG = "AudioAsset";

    private final ArrayList<AudioDispatcher> audioList;
    protected int playIndex = 0;
    protected final NativeAudio owner;
    protected AudioCompletionListener completionListener;
    protected String assetId;
    protected Handler currentTimeHandler;
    protected Runnable currentTimeRunnable;
    protected static final int FADE_DELAY_MS = 80; // Delay between fade steps in milliseconds

    protected ScheduledExecutorService fadeExecutor;
    protected ScheduledFuture<?> fadeTask;

    protected Map<String, Boolean> dispatchedCompleteMap = new HashMap<>();

    protected enum FadeState {
        NONE,
        FADE_IN,
        FADE_OUT,
        FADE_TO
    }

    protected FadeState fadeState = FadeState.NONE;

    protected final float zeroVolume = 0.001f; // Minimum volume to avoid zero for exponential fade
    protected final float maxVolume = 1.0f; // Maximum volume

    AudioAsset(NativeAudio owner, String assetId, AssetFileDescriptor assetFileDescriptor, int audioChannelNum, float volume)
        throws Exception {
        audioList = new ArrayList<>();
        this.owner = owner;
        this.assetId = assetId;
        this.fadeExecutor = Executors.newSingleThreadScheduledExecutor();

        if (audioChannelNum < 0) {
            audioChannelNum = 1;
        }

        for (int x = 0; x < audioChannelNum; x++) {
            AudioDispatcher audioDispatcher = new AudioDispatcher(assetFileDescriptor, volume);
            audioList.add(audioDispatcher);
            if (audioChannelNum == 1) audioDispatcher.setOwner(this);
        }
    }

    public void dispatchComplete() {
        if (dispatchedCompleteMap.getOrDefault(this.assetId, false)) {
            return;
        }
        this.owner.dispatchComplete(this.assetId);
        dispatchedCompleteMap.put(this.assetId, true);
    }

    public void play(double time, float volume) throws Exception {
        AudioDispatcher audio = audioList.get(playIndex);
        if (audio != null) {
            cancelFade();
            audio.play(time);
            audio.setVolume(volume);
            playIndex++;
            playIndex = playIndex % audioList.size();
            Log.d(TAG, "Starting timer from play"); // Debug log
            startCurrentTimeUpdates(); // Make sure this is called
        } else {
            throw new Exception("AudioDispatcher is null");
        }
    }

    public double getDuration() {
        if (audioList.size() != 1) return 0;

        AudioDispatcher audio = audioList.get(playIndex);

        if (audio != null) {
            return audio.getDuration();
        }
        return 0;
    }

    public void setCurrentPosition(double time) {
        if (audioList.size() != 1) return;

        AudioDispatcher audio = audioList.get(playIndex);

        if (audio != null) {
            audio.setCurrentPosition(time);
        }
    }

    public double getCurrentPosition() {
        if (audioList.size() != 1) return 0;

        AudioDispatcher audio = audioList.get(playIndex);

        if (audio != null) {
            return audio.getCurrentPosition();
        }
        return 0;
    }

    public boolean pause() throws Exception {
        stopCurrentTimeUpdates(); // Stop updates when pausing
        boolean wasPlaying = false;

        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);
            if (audio == null) {
                continue;
            }
            cancelFade();
            wasPlaying |= audio.pause();
        }

        return wasPlaying;
    }

    public void resume() throws Exception {
        if (!audioList.isEmpty()) {
            AudioDispatcher audio = audioList.get(0);
            if (audio != null) {
                audio.resume();
                Log.d(TAG, "Starting timer from resume"); // Debug log
                startCurrentTimeUpdates(); // Make sure this is called
            } else {
                throw new Exception("AudioDispatcher is null");
            }
        }
    }

    public void stop() throws Exception {
        stopCurrentTimeUpdates(); // Stop updates when stopping
        dispatchComplete();
        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);

            if (audio != null) {
                cancelFade();
                audio.stop();
            } else {
                throw new Exception("AudioDispatcher is null");
            }
        }
    }

    public void loop() throws Exception {
        AudioDispatcher audio = audioList.get(playIndex);
        if (audio != null) {
            audio.loop();
            playIndex++;
            playIndex = playIndex % audioList.size();
            startCurrentTimeUpdates(); // Add timer start
        } else {
            throw new Exception("AudioDispatcher is null");
        }
    }

    public void unload() throws Exception {
        this.stop();

        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);

            if (audio != null) {
                audio.unload();
            } else {
                throw new Exception("AudioDispatcher is null");
            }
        }

        audioList.clear();
        fadeExecutor.shutdown();
    }

    public void setVolume(float volume, double duration) throws Exception {
        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);

            cancelFade();
            if (audio != null) {
                if (isPlaying() && duration > 0) {
                    fadeTo(audio, duration, volume);
                } else {
                    audio.setVolume(volume);
                }
            } else {
                throw new Exception("AudioDispatcher is null");
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    public void setRate(float rate) throws Exception {
        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);
            if (audio != null) {
                audio.setRate(rate);
            }
        }
    }

    public boolean isPlaying() throws Exception {
        for (AudioDispatcher ad : audioList) {
            if (ad.isPlaying()) return true;
        }
        return false;
    }

    public void setCompletionListener(AudioCompletionListener listener) {
        this.completionListener = listener;
    }

    protected void notifyCompletion() {
        if (completionListener != null) {
            completionListener.onCompletion(this.assetId);
        }
    }

    protected String getAssetId() {
        return assetId;
    }

    public void setCurrentTime(double time) throws Exception {
        owner
            .getActivity()
            .runOnUiThread(
                new Runnable() {
                    @Override
                    public void run() {
                        if (audioList.size() != 1) {
                            return;
                        }
                        AudioDispatcher audio = audioList.get(playIndex);
                        if (audio != null) {
                            audio.setCurrentPosition(time);
                        }
                    }
                }
            );
    }

    protected void startCurrentTimeUpdates() {
        Log.d(TAG, "Starting timer updates");
        if (currentTimeHandler == null) {
            currentTimeHandler = new Handler(Looper.getMainLooper());
        }
        // Reset completion status for this assetId
        dispatchedCompleteMap.put(assetId, false);

        // Add small delay to let audio start playing
        currentTimeHandler.postDelayed(
            new Runnable() {
                @Override
                public void run() {
                    startTimeUpdateLoop();
                }
            },
            100
        ); // 100ms delay
    }

    private void startTimeUpdateLoop() {
        currentTimeRunnable = new Runnable() {
            @Override
            public void run() {
                AudioDispatcher audio = null;
                try {
                    audio = audioList.get(playIndex);
                } catch (Exception e) {
                    Log.v(TAG, "Audio dispatcher does not exist at index " + playIndex);
                }
                if (audio == null) {
                    Log.d(TAG, "Audio dispatcher does not exist - aborting timer update");
                    return;
                }

                try {
                    if (audio != null && audio.isPlaying()) {
                        double currentTime = getCurrentPosition();
                        Log.v(TAG, "Play timer update: currentTime = " + currentTime);
                        owner.notifyCurrentTime(assetId, currentTime);
                        currentTimeHandler.postDelayed(this, 100);
                    } else {
                        Log.d(TAG, "Stopping play timer - not playing");
                        stopCurrentTimeUpdates();
                        dispatchComplete();
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error getting current time", e);
                    stopCurrentTimeUpdates();
                }
            }
        };
        currentTimeHandler.post(currentTimeRunnable);
    }

    void stopCurrentTimeUpdates() {
        Log.d(TAG, "Stopping play timer updates");
        if (currentTimeHandler != null && currentTimeRunnable != null) {
            currentTimeHandler.removeCallbacks(currentTimeRunnable);
            currentTimeHandler = null;
            currentTimeRunnable = null;
        }
    }

    public void playWithFadeIn(double time, float volume, double fadeInDurationMs) throws Exception {
        AudioDispatcher audio = audioList.get(playIndex);
        if (audio != null) {
            audio.setVolume(0);
            audio.play(time);
            fadeIn(audio, fadeInDurationMs, volume);
            startCurrentTimeUpdates();
        }
    }

    private void fadeIn(final AudioDispatcher audio, double fadeInDurationMs, float targetVolume) {
        cancelFade();
        fadeState = FadeState.FADE_IN;

        final int steps = Math.max(1, (int) (fadeInDurationMs / FADE_DELAY_MS));
        final float fadeStep = targetVolume / steps;

        Log.d(
            TAG,
            "Beginning fade in at time " +
            getCurrentPosition() +
            " over " +
            (fadeInDurationMs / 1000.0) +
            "s to target volume " +
            targetVolume +
            " in " +
            steps +
            " steps (step duration: " +
            (FADE_DELAY_MS / 1000.0) +
            "s"
        );

        fadeTask = fadeExecutor.scheduleWithFixedDelay(
            new Runnable() {
                float currentVolume = 0;

                @Override
                public void run() {
                    if (fadeState != FadeState.FADE_IN || currentVolume >= targetVolume) {
                        fadeState = FadeState.NONE;
                        cancelFade();
                        Log.d(TAG, "Fade in complete at time " + getCurrentPosition());
                        return;
                    }
                    final float previousCurrentVolume = currentVolume;
                    currentVolume += fadeStep;
                    try {
                        final float resolvedTargetVolume = Math.min(Math.max(currentVolume, 0), targetVolume);
                        Log.v(
                            TAG,
                            "Fade in step: from " + previousCurrentVolume + " to " + currentVolume + " to target " + resolvedTargetVolume
                        );
                        if (audio != null) audio.setVolume(resolvedTargetVolume);
                    } catch (Exception e) {
                        Log.e(TAG, "Error during fade in", e);
                        cancelFade();
                    }
                }
            },
            0,
            FADE_DELAY_MS,
            TimeUnit.MILLISECONDS
        );
    }

    public void stopWithFade(double fadeOutDurationMs) throws Exception {
        AudioDispatcher audio = audioList.get(playIndex);
        if (audio != null && audio.isPlaying()) {
            cancelFade();
            fadeOut(audio, fadeOutDurationMs);
        }
    }

    private void fadeOut(final AudioDispatcher audio, double fadeOutDurationMs) {
        cancelFade();
        fadeState = FadeState.FADE_OUT;

        if (audio == null) return;

        final int steps = Math.max(1, (int) (fadeOutDurationMs / FADE_DELAY_MS));
        final float initialVolume = audio.getVolume();
        final float fadeStep = initialVolume / steps;

        Log.d(
            TAG,
            "Beginning fade out from volume " +
            initialVolume +
            " at time " +
            getCurrentPosition() +
            " over " +
            (fadeOutDurationMs / 1000.0) +
            "s in " +
            steps +
            " steps (step duration: " +
            (FADE_DELAY_MS / 1000.0) +
            "s)"
        );

        fadeTask = fadeExecutor.scheduleWithFixedDelay(
            new Runnable() {
                float currentVolume = initialVolume;

                @Override
                public void run() {
                    if (fadeState != FadeState.FADE_OUT || currentVolume <= 0) {
                        fadeState = FadeState.NONE;
                        stopAudio(audio);
                        cancelFade();
                        Log.d(TAG, "Fade out complete at time " + getCurrentPosition());
                        return;
                    }
                    final float previousCurrentVolume = currentVolume;
                    currentVolume -= fadeStep;
                    try {
                        final float thisTargetVolume = Math.max(currentVolume, 0);
                        Log.v(
                            TAG,
                            "Fade out step: from " + previousCurrentVolume + " to " + currentVolume + " to target " + thisTargetVolume
                        );
                        if (audio != null) audio.setVolume(thisTargetVolume);
                    } catch (Exception e) {
                        Log.e(TAG, "Error during fade out", e);
                        cancelFade();
                    }
                }
            },
            0,
            FADE_DELAY_MS,
            TimeUnit.MILLISECONDS
        );
    }

    private void fadeTo(final AudioDispatcher audio, double fadeDurationMs, float targetVolume) {
        cancelFade();
        fadeState = FadeState.FADE_TO;

        if (audio == null) return;

        final int steps = Math.max(1, (int) (fadeDurationMs / FADE_DELAY_MS));
        final float minVolume = zeroVolume;
        final float initialVolume = Math.max(audio.getVolume(), minVolume);
        final float finalTargetVolume = Math.max(targetVolume, minVolume);

        // Calculate exponential ratio for perceptual fade
        final float safeInitialVolume = Math.max(initialVolume, zeroVolume);
        final double ratio = Math.pow(finalTargetVolume / safeInitialVolume, 1.0 / steps);

        Log.d(
            TAG,
            "Beginning exponential fade from volume " +
            initialVolume +
            " to " +
            finalTargetVolume +
            " over " +
            (fadeDurationMs / 1000.0) +
            "s in " +
            steps +
            " steps (step duration: " +
            (FADE_DELAY_MS / 1000.0) +
            "s)"
        );

        fadeTask = fadeExecutor.scheduleWithFixedDelay(
            new Runnable() {
                int currentStep = 0;
                float currentVolume = initialVolume;

                @Override
                public void run() {
                    if ((audio != null && fadeState != FadeState.FADE_TO) || !audio.isPlaying() || currentStep >= steps) {
                        fadeState = FadeState.NONE;
                        cancelFade();
                        Log.d(TAG, "Fade to complete at time " + getCurrentPosition());
                        return;
                    }

                    try {
                        currentVolume *= (float) ratio;
                        currentVolume = Math.min(Math.max(currentVolume, minVolume), maxVolume); // Clamp between minVolume and maxVolume
                        if (audio != null) audio.setVolume(currentVolume);
                        Log.v(TAG, "Fade to step " + currentStep + ": volume set to " + currentVolume);
                        currentStep++;
                    } catch (Exception e) {
                        Log.e(TAG, "Error during fade to", e);
                        cancelFade();
                    }
                }
            },
            0,
            FADE_DELAY_MS,
            TimeUnit.MILLISECONDS
        );
    }

    /**
     * Cancels the fade task if it is running.
     */
    private void cancelFade() {
        if (fadeTask != null && !fadeTask.isCancelled()) {
            fadeTask.cancel(true);
        }
        fadeState = FadeState.NONE;
        fadeTask = null;
    }

    private void stopAudio(final AudioDispatcher audio) {
        if (audio != null) {
            try {
                audio.setVolume(0);
                stop();
                cancelFade();
            } catch (Exception e) {
                Log.e(TAG, "Error stopping after fade", e);
            }
        }
    }
}
